File: src/security/core/ScanEngine.ts
import { SecurityViolation, ScanContext } from '../types/SecurityViolation';

export interface SecurityPattern {
  id: string;
  category: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  pattern: RegExp;
  description: string;
  recommendation: string;
  cweReference?: string;
}

export interface ViolationDetector {
  readonly category: string;
  readonly severity: ViolationSeverity;
  
  detect(content: string, context: ScanContext): Promise<SecurityViolation[]>;
  getPatterns(): SecurityPattern[];
  generateFix(violation: SecurityViolation): string;
}

export type ViolationSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';

// OWASP LLM Top 10 Pattern Definitions
export const OWASP_LLM_PATTERNS: SecurityPattern[] = [
  // LLM01: Prompt Injection
  {
    id: 'LLM01-001',
    category: 'LLM01',
    severity: 'HIGH',
    pattern: /(?:prompt|input)\s*=\s*["'`][^"'`]*?(?:ignore|forget|disregard|system|admin|root)/i,
    description: 'Potential prompt injection vulnerability detected',
    recommendation: 'Validate and sanitize all user inputs before including in prompts',
    cweReference: 'CWE-74'
  },
  
  // LLM06: Sensitive Information Disclosure
  {
    id: 'LLM06-001',
    category: 'LLM06',
    severity: 'CRITICAL',
    pattern: /(?:api_key|apikey|secret|token|password|pwd)\s*=\s*["'`][a-zA-Z0-9_-]{10,}["'`]/i,
    description: 'Hardcoded sensitive information detected',
    recommendation: 'Move sensitive data to environment variables or secure configuration',
    cweReference: 'CWE-798'
  },
  
  // LLM06: Additional API Key patterns
  {
    id: 'LLM06-002',
    category: 'LLM06',
    severity: 'CRITICAL',
    pattern: /(?:sk-|pk_|sk_|AIza)[a-zA-Z0-9_-]{20,}/,
    description: 'Potential API key or token detected',
    recommendation: 'Remove hardcoded credentials and use secure key management',
    cweReference: 'CWE-798'
  },
  
  // LLM02: Insecure Output Handling
  {
    id: 'LLM02-001',
    category: 'LLM02',
    severity: 'MEDIUM',
    pattern: /(?:dangerouslySetInnerHTML|innerHTML|outerHTML)\s*=\s*[^;]+;/i,
    description: 'Direct HTML assignment without sanitization',
    recommendation: 'Use proper output sanitization libraries like DOMPurify',
    cweReference: 'CWE-79'
  },
  
  // Agent OS Coding Standards
  {
    id: 'AOS-001',
    category: 'CODING_STANDARDS',
    severity: 'HIGH',
    pattern: /console\.(log|debug|info|warn|error)\s*\([^)]*["'`][^"'`]*?(?:password|secret|token|key)/i,
    description: 'Sensitive information logged to console',
    recommendation: 'Remove sensitive data from console logs',
    cweReference: 'CWE-532'
  },
  
  {
    id: 'AOS-002',
    category: 'CODING_STANDARDS',
    severity: 'MEDIUM',
    pattern: /catch\s*\(\s*\)\s*\{\s*\}/,
    description: 'Empty catch block detected',
    recommendation: 'Add proper error handling or logging in catch blocks',
    cweReference: 'CWE-390'
  }
];

class ScanEngine {
  private detectors: Map<string, ViolationDetector>;
  private patterns: SecurityPattern[];

  constructor() {
    this.detectors = new Map();
    this.patterns = OWASP_LLM_PATTERNS;
    this.initializeDetectors();
  }

  private initializeDetectors(): void {
    // Initialize OWASP LLM detectors
    this.detectors.set('LLM01', new PromptInjectionDetector());
    this.detectors.set('LLM02', new InsecureOutputDetector());
    this.detectors.set('LLM06', new SensitiveInfoDetector());
    this.detectors.set('CODING_STANDARDS', new CodingStandardsDetector());
  }

  async scanProject(projectPath: string, context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    
    // For now, we'll simulate scanning files
    // In a real implementation, this would:
    // 1. Read all files in the project
    // 2. Apply each detector to each file
    // 3. Aggregate results
    
    const mockViolations = await this.generateMockViolations(context);
    violations.push(...mockViolations);
    
    return violations;
  }

  async scanFile(filePath: string, content: string, context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    
    // Apply all detectors to the file content
    for (const [category, detector] of this.detectors) {
      const fileViolations = await detector.detect(content, context);
      violations.push(...fileViolations);
    }
    
    return violations;
  }

  async startRealTimeScan(projectPath: string): Promise<AsyncIterable<SecurityViolation>> {
    // This would implement real-time file watching
    // For now, return an empty async iterable
    return (async function*() {
      // Yield violations as they're detected
    })();
  }

  private async generateMockViolations(context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    
    // Generate mock violations based on agent source
    if (context.agentSource === 'windsurf') {
      violations.push(this.createViolation(
        'LLM06-001',
        'CRITICAL',
        'LLM06',
        'Hardcoded API key detected',
        'API key found in configuration file',
        'src/config/api.ts',
        15,
        'const API_KEY = "sk-1234567890abcdef";',
        'Move API key to environment variables',
        'CWE-798',
        context.agentSource
      ));
    }
    
    if (context.agentSource === 'anti-gravity') {
      violations.push(this.createViolation(
        'LLM02-001',
        'MEDIUM',
        'LLM02',
        'Insecure output handling',
        'Raw HTML output without sanitization',
        'src/components/Output.tsx',
        42,
        'div.innerHTML = userInput;',
        'Use DOMPurify or similar for sanitization',
        'CWE-79',
        context.agentSource
      ));
    }
    
    return violations;
  }

  private createViolation(
    patternId: string,
    severity: ViolationSeverity,
    category: string,
    title: string,
    description: string,
    file: string,
    line: number,
    codeSnippet: string,
    recommendation: string,
    cweReference?: string,
    agentSource?: string
  ): SecurityViolation {
    return {
      id: `violation_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      severity,
      category,
      title,
      description,
      file,
      line,
      codeSnippet,
      recommendation,
      cweReference,
      agentSource,
      status: 'OPEN',
      discoveredAt: new Date()
    };
  }

  getPatterns(): SecurityPattern[] {
    return [...this.patterns];
  }

  addPattern(pattern: SecurityPattern): void {
    this.patterns.push(pattern);
  }

  removePattern(patternId: string): void {
    this.patterns = this.patterns.filter(p => p.id !== patternId);
  }
}

// Individual Detector Implementations

class PromptInjectionDetector implements ViolationDetector {
  readonly category = 'LLM01';
  readonly severity = 'HIGH';

  async detect(content: string, context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    const patterns = OWASP_LLM_PATTERNS.filter(p => p.category === 'LLM01');
    
    for (const pattern of patterns) {
      const matches = content.matchAll(pattern.pattern);
      for (const match of matches) {
        const lines = content.split('\n');
        const lineNumber = lines.findIndex(line => line.includes(match[0])) + 1;
        
        violations.push({
          id: `violation_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
          severity: pattern.severity,
          category: pattern.category,
          title: 'Prompt Injection Vulnerability',
          description: pattern.description,
          file: context.projectPath, // Would be actual file path
          line: lineNumber,
          codeSnippet: match[0],
          recommendation: pattern.recommendation,
          cweReference: pattern.cweReference,
          agentSource: context.agentSource,
          status: 'OPEN',
          discoveredAt: new Date()
        });
      }
    }
    
    return violations;
  }

  getPatterns(): SecurityPattern[] {
    return OWASP_LLM_PATTERNS.filter(p => p.category === 'LLM01');
  }

  generateFix(violation: SecurityViolation): string {
    return `// TODO: Implement input validation for prompt injection prevention
const sanitizedInput = validateAndSanitizeInput(userInput);
const prompt = \`System: You are a helpful assistant.\\nUser: \${sanitizedInput}\`;`;
  }
}

class InsecureOutputDetector implements ViolationDetector {
  readonly category = 'LLM02';
  readonly severity = 'MEDIUM';

  async detect(content: string, context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    const patterns = OWASP_LLM_PATTERNS.filter(p => p.category === 'LLM02');
    
    for (const pattern of patterns) {
      const matches = content.matchAll(pattern.pattern);
      for (const match of matches) {
        const lines = content.split('\n');
        const lineNumber = lines.findIndex(line => line.includes(match[0])) + 1;
        
        violations.push({
          id: `violation_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
          severity: pattern.severity,
          category: pattern.category,
          title: 'Insecure Output Handling',
          description: pattern.description,
          file: context.projectPath,
          line: lineNumber,
          codeSnippet: match[0],
          recommendation: pattern.recommendation,
          cweReference: pattern.cweReference,
          agentSource: context.agentSource,
          status: 'OPEN',
          discoveredAt: new Date()
        });
      }
    }
    
    return violations;
  }

  getPatterns(): SecurityPattern[] {
    return OWASP_LLM_PATTERNS.filter(p => p.category === 'LLM02');
  }

  generateFix(violation: SecurityViolation): string {
    return `import DOMPurify from 'dompurify';
// Sanitize LLM output before rendering
const sanitizedOutput = DOMPurify.sanitize(llmResponse);
div.innerHTML = sanitizedOutput;`;
  }
}

class SensitiveInfoDetector implements ViolationDetector {
  readonly category = 'LLM06';
  readonly severity = 'CRITICAL';

  async detect(content: string, context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    const patterns = OWASP_LLM_PATTERNS.filter(p => p.category === 'LLM06');
    
    for (const pattern of patterns) {
      const matches = content.matchAll(pattern.pattern);
      for (const match of matches) {
        const lines = content.split('\n');
        const lineNumber = lines.findIndex(line => line.includes(match[0])) + 1;
        
        violations.push({
          id: `violation_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
          severity: pattern.severity,
          category: pattern.category,
          title: 'Sensitive Information Disclosure',
          description: pattern.description,
          file: context.projectPath,
          line: lineNumber,
          codeSnippet: match[0],
          recommendation: pattern.recommendation,
          cweReference: pattern.cweReference,
          agentSource: context.agentSource,
          status: 'OPEN',
          discoveredAt: new Date()
        });
      }
    }
    
    return violations;
  }

  getPatterns(): SecurityPattern[] {
    return OWASP_LLM_PATTERNS.filter(p => p.category === 'LLM06');
  }

  generateFix(violation: SecurityViolation): string {
    return `// Move sensitive data to environment variables
const API_KEY = process.env.API_KEY;
// or use a secure configuration manager
const config = await secureConfigManager.getSecret('api-key');`;
  }
}

class CodingStandardsDetector implements ViolationDetector {
  readonly category = 'CODING_STANDARDS';
  readonly severity = 'MEDIUM';

  async detect(content: string, context: ScanContext): Promise<SecurityViolation[]> {
    const violations: SecurityViolation[] = [];
    const patterns = OWASP_LLM_PATTERNS.filter(p => p.category === 'CODING_STANDARDS');
    
    for (const pattern of patterns) {
      const matches = content.matchAll(pattern.pattern);
      for (const match of matches) {
        const lines = content.split('\n');
        const lineNumber = lines.findIndex(line => line.includes(match[0])) + 1;
        
        violations.push({
          id: `violation_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
          severity: pattern.severity,
          category: pattern.category,
          title: 'Coding Standards Violation',
          description: pattern.description,
          file: context.projectPath,
          line: lineNumber,
          codeSnippet: match[0],
          recommendation: pattern.recommendation,
          cweReference: pattern.cweReference,
          agentSource: context.agentSource,
          status: 'OPEN',
          discoveredAt: new Date()
        });
      }
    }
    
    return violations;
  }

  getPatterns(): SecurityPattern[] {
    return OWASP_LLM_PATTERNS.filter(p => p.category === 'CODING_STANDARDS');
  }

  generateFix(violation: SecurityViolation): string {
    if (violation.codeSnippet.includes('console.log')) {
      return `// Remove sensitive data from logs
console.log('Processing request for user:', userId); // Safe
// console.log('User password:', password); // UNSAFE - remove this`;
    }
    
    if (violation.codeSnippet.includes('catch')) {
      return `catch (error) {
  // Add proper error handling
  logger.error('Operation failed', { error, context });
  throw new Error('Operation could not be completed');
}`;
    }
    
    return '// TODO: Fix coding standards violation';
  }
}

export default ScanEngine;

File: src/security_py/core/scan_engine.py
"""
Deterministic Layer - Pattern-Based Security Scanning

Uses OWASP LLM Top 10 patterns as Python dataclasses with compiled RegEx
for high-speed scanning of Python source code.
"""

import re
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Iterator

from ..types.violations import (
    SecurityViolation,
    Severity,
    ViolationStatus,
    ScanContext,
)


@dataclass(frozen=True)
class SecurityPattern:
    """OWASP LLM security pattern definition."""
    id: str
    category: str
    severity: Severity
    pattern: re.Pattern
    description: str
    recommendation: str
    cwe_reference: str = ""

    @classmethod
    def compile(
        cls,
        id: str,
        category: str,
        severity: Severity,
        pattern: str,
        description: str,
        recommendation: str,
        cwe_reference: str = "",
        flags: int = re.IGNORECASE | re.MULTILINE,
    ) -> "SecurityPattern":
        """Create a SecurityPattern with compiled regex."""
        return cls(
            id=id,
            category=category,
            severity=severity,
            pattern=re.compile(pattern, flags),
            description=description,
            recommendation=recommendation,
            cwe_reference=cwe_reference,
        )


# OWASP LLM Top 10 Patterns - Python-Focused
OWASP_LLM_PATTERNS: tuple[SecurityPattern, ...] = (
    # LLM01: Prompt Injection
    SecurityPattern.compile(
        id="LLM01-001",
        category="LLM01",
        severity=Severity.HIGH,
        pattern=r'(?:prompt|input|query)\s*=\s*["\'][^"\']*?(?:ignore|forget|disregard|system|admin|root)',
        description="Potential prompt injection vulnerability detected",
        recommendation="Validate and sanitize all user inputs before including in prompts",
        cwe_reference="CWE-74",
    ),
    SecurityPattern.compile(
        id="LLM01-002",
        category="LLM01",
        severity=Severity.HIGH,
        pattern=r'f["\'].*\{.*input\(\).*\}.*["\']',
        description="Unsanitized user input in f-string prompt",
        recommendation="Sanitize input() values before embedding in prompts",
        cwe_reference="CWE-74",
    ),
    # LLM06: Sensitive Information Disclosure
    SecurityPattern.compile(
        id="LLM06-001",
        category="LLM06",
        severity=Severity.CRITICAL,
        pattern=r'(?:api_key|apikey|secret|token|password|pwd)\s*=\s*["\'][a-zA-Z0-9_\-]{10,}["\']',
        description="Hardcoded sensitive information detected",
        recommendation="Move sensitive data to environment variables using os.environ",
        cwe_reference="CWE-798",
    ),
    SecurityPattern.compile(
        id="LLM06-002",
        category="LLM06",
        severity=Severity.CRITICAL,
        pattern=r'(?:sk-|pk_|sk_|AIza|ghp_|gho_|glpat-)[a-zA-Z0-9_\-]{20,}',
        description="Potential API key or token detected (OpenAI, Stripe, GitHub, Google, GitLab)",
        recommendation="Remove hardcoded credentials and use secure key management",
        cwe_reference="CWE-798",
    ),
    SecurityPattern.compile(
        id="LLM06-003",
        category="LLM06",
        severity=Severity.HIGH,
        pattern=r'print\s*\([^)]*(?:password|secret|token|key|credential)',
        description="Sensitive information logged to stdout via print()",
        recommendation="Remove sensitive data from print statements",
        cwe_reference="CWE-532",
    ),
    # LLM02: Insecure Output Handling
    SecurityPattern.compile(
        id="LLM02-001",
        category="LLM02",
        severity=Severity.CRITICAL,
        pattern=r'eval\s*\([^)]+\)',
        description="Use of eval() - arbitrary code execution risk",
        recommendation="Replace eval() with ast.literal_eval() or safer alternatives",
        cwe_reference="CWE-95",
    ),
    SecurityPattern.compile(
        id="LLM02-002",
        category="LLM02",
        severity=Severity.CRITICAL,
        pattern=r'exec\s*\([^)]+\)',
        description="Use of exec() - arbitrary code execution risk",
        recommendation="Avoid exec(); use structured data processing instead",
        cwe_reference="CWE-95",
    ),
    SecurityPattern.compile(
        id="LLM02-003",
        category="LLM02",
        severity=Severity.HIGH,
        pattern=r'__import__\s*\([^)]+\)',
        description="Dynamic import with __import__() - code injection risk",
        recommendation="Use importlib.import_module() with validated module names",
        cwe_reference="CWE-95",
    ),
    # Shell/Command Injection
    SecurityPattern.compile(
        id="CMD-001",
        category="COMMAND_INJECTION",
        severity=Severity.CRITICAL,
        pattern=r'os\.system\s*\([^)]+\)',
        description="Use of os.system() - command injection risk",
        recommendation="Use subprocess.run() with shell=False and argument list",
        cwe_reference="CWE-78",
    ),
    SecurityPattern.compile(
        id="CMD-002",
        category="COMMAND_INJECTION",
        severity=Severity.CRITICAL,
        pattern=r'subprocess\.(?:call|run|Popen)\s*\([^)]*shell\s*=\s*True',
        description="Subprocess with shell=True - command injection risk",
        recommendation="Use shell=False with argument list instead",
        cwe_reference="CWE-78",
    ),
    SecurityPattern.compile(
        id="CMD-003",
        category="COMMAND_INJECTION",
        severity=Severity.HIGH,
        pattern=r'subprocess\.(?:call|run|Popen)\s*\(\s*f["\']',
        description="Subprocess with f-string - potential command injection",
        recommendation="Use argument list instead of string interpolation",
        cwe_reference="CWE-78",
    ),
    # Coding Standards
    SecurityPattern.compile(
        id="PY-001",
        category="CODING_STANDARDS",
        severity=Severity.MEDIUM,
        pattern=r'except\s*:\s*(?:pass|\.\.\.)',
        description="Bare except with pass - silent failure",
        recommendation="Log exceptions and handle specific exception types",
        cwe_reference="CWE-390",
    ),
    SecurityPattern.compile(
        id="PY-002",
        category="CODING_STANDARDS",
        severity=Severity.MEDIUM,
        pattern=r'pickle\.(?:load|loads)\s*\(',
        description="Pickle deserialization - arbitrary code execution risk",
        recommendation="Use JSON or other safe serialization formats",
        cwe_reference="CWE-502",
    ),
    SecurityPattern.compile(
        id="PY-003",
        category="CODING_STANDARDS",
        severity=Severity.HIGH,
        pattern=r'yaml\.(?:load|unsafe_load)\s*\([^)]*\)',
        description="Unsafe YAML loading - arbitrary code execution risk",
        recommendation="Use yaml.safe_load() instead",
        cwe_reference="CWE-502",
    ),
    # SQL Injection
    SecurityPattern.compile(
        id="SQL-001",
        category="SQL_INJECTION",
        severity=Severity.CRITICAL,
        pattern=r'(?:execute|cursor\.execute)\s*\(\s*f["\'].*(?:SELECT|INSERT|UPDATE|DELETE)',
        description="SQL query with f-string - SQL injection risk",
        recommendation="Use parameterized queries with ? or %s placeholders",
        cwe_reference="CWE-89",
    ),
    SecurityPattern.compile(
        id="SQL-002",
        category="SQL_INJECTION",
        severity=Severity.CRITICAL,
        pattern=r'(?:execute|cursor\.execute)\s*\([^)]*%\s*\(',
        description="SQL query with string formatting - SQL injection risk",
        recommendation="Use parameterized queries with ? or %s placeholders",
        cwe_reference="CWE-89",
    ),
)


class ScanEngine:
    """
    Deterministic Layer: High-speed pattern matching for Python source code.
    
    Uses compiled regex patterns from OWASP LLM Top 10 to detect
    security vulnerabilities in Python files.
    """

    def __init__(self, patterns: tuple[SecurityPattern, ...] = OWASP_LLM_PATTERNS):
        self._patterns = patterns

    @property
    def patterns(self) -> tuple[SecurityPattern, ...]:
        """Get all registered patterns."""
        return self._patterns

    def scan_content(
        self,
        content: str,
        file_path: str,
        context: ScanContext,
    ) -> list[SecurityViolation]:
        """
        Scan source code content for pattern matches.
        
        Args:
            content: Source code to scan
            file_path: Path to the file being scanned
            context: Scan context with metadata
            
        Returns:
            List of SecurityViolation objects for each match
        """
        violations: list[SecurityViolation] = []
        lines = content.splitlines()

        for pattern in self._patterns:
            for match in pattern.pattern.finditer(content):
                # Calculate line number from match position
                line_num = content[:match.start()].count('\n') + 1
                
                # Get the matched line as code snippet
                if 0 < line_num <= len(lines):
                    code_snippet = lines[line_num - 1].strip()
                else:
                    code_snippet = match.group(0)[:80]

                violation = SecurityViolation(
                    severity=pattern.severity,
                    category=pattern.category,
                    title=f"{pattern.category}: {pattern.id}",
                    description=pattern.description,
                    file=file_path,
                    line=line_num,
                    code_snippet=code_snippet,
                    recommendation=pattern.recommendation,
                    cwe_reference=pattern.cwe_reference,
                    agent_source=context.agent_source,
                )
                violations.append(violation)

        return violations

    def scan_file(self, file_path: Path, context: ScanContext) -> list[SecurityViolation]:
        """
        Scan a single Python file for security violations.
        
        Args:
            file_path: Path to the Python file
            context: Scan context with metadata
            
        Returns:
            List of SecurityViolation objects
        """
        try:
            content = file_path.read_text(encoding="utf-8")
            return self.scan_content(content, str(file_path), context)
        except (OSError, UnicodeDecodeError) as e:
            # Return a violation indicating scan failure
            return [
                SecurityViolation(
                    severity=Severity.LOW,
                    category="SCAN_ERROR",
                    title="File scan failed",
                    description=f"Could not scan file: {e}",
                    file=str(file_path),
                    line=0,
                    code_snippet="",
                    recommendation="Check file encoding and permissions",
                )
            ]

    def scan_directory(
        self,
        directory: Path,
        context: ScanContext,
        extensions: tuple[str, ...] = (".py",),
    ) -> Iterator[SecurityViolation]:
        """
        Recursively scan a directory for Python files.
        
        Args:
            directory: Root directory to scan
            context: Scan context with metadata
            extensions: File extensions to include
            
        Yields:
            SecurityViolation objects as they are found
        """
        for file_path in directory.rglob("*"):
            if file_path.suffix in extensions and file_path.is_file():
                # Skip virtual environments and common non-source dirs
                if any(
                    part in file_path.parts
                    for part in ("venv", ".venv", "node_modules", "__pycache__", ".git")
                ):
                    continue
                
                for violation in self.scan_file(file_path, context):
                    yield violation

    def add_pattern(self, pattern: SecurityPattern) -> None:
        """Add a custom security pattern."""
        self._patterns = (*self._patterns, pattern)

    def remove_pattern(self, pattern_id: str) -> bool:
        """Remove a pattern by ID. Returns True if removed."""
        original_len = len(self._patterns)
        self._patterns = tuple(p for p in self._patterns if p.id != pattern_id)
        return len(self._patterns) < original_len
